File: frontend/js/sec_leads.js
/**
 * sec_leads.js — SEC 10-K Patent Importance Analysis tab.
 *
 * Loads reports from /api/sec-leads and renders the analysis table.
 * Memo and Letter buttons open a modal popup with generated documents.
 */

import {
  apiGet, showStatus, escHtml,
  enableTableSorting, stampOriginalOrder,
} from './app.js';

// ── DOM references ─────────────────────────────────────────────────

const datePicker    = document.getElementById('sec-date-picker');
const loadBtn       = document.getElementById('sec-load-btn');
const statsGrid     = document.getElementById('sec-stats');
const filingsCount  = document.getElementById('sec-filings-count');
const score5Count   = document.getElementById('sec-score5-count');
const score7Count   = document.getElementById('sec-score7-count');
const reportDate    = document.getElementById('sec-report-date');
const resultsArea   = document.getElementById('sec-results');
const resultsCount  = document.getElementById('sec-results-count');
const tableBody     = document.getElementById('sec-table-body');
const tableEl       = document.getElementById('sec-table');
const statusEl      = document.getElementById('sec-status');

// Modal
const modal         = document.getElementById('sec-doc-modal');
const modalTitle    = document.getElementById('sec-modal-title');
const modalBody     = document.getElementById('sec-modal-body');
const modalCopyBtn  = document.getElementById('sec-modal-copy-btn');
const modalCloseBtn = document.getElementById('sec-modal-close-btn');
const modalBackdrop = modal?.querySelector('.modal-backdrop');

// ── Score badge colors ─────────────────────────────────────────────

const SCORE_COLORS = {
  10: '#c0392b', 9: '#c0392b',
  8: '#e74c3c',
  7: '#e67e22',
  6: '#f39c12',
  5: '#95a5a6',
};

function scoreColor(score) {
  return SCORE_COLORS[score] || '#bdc3c7';
}

// ── State ──────────────────────────────────────────────────────────

let loaded = false;
let currentDate = '';

// ── Load report dates into picker ──────────────────────────────────

async function loadReportDates() {
  try {
    const data = await apiGet('/api/sec-leads/reports');
    const reports = data.reports || [];
    datePicker.innerHTML = '';

    if (reports.length === 0) {
      datePicker.innerHTML = '<option value="">No reports available</option>';
      return;
    }

    reports.forEach((r, i) => {
      const opt = document.createElement('option');
      opt.value = r.analysis_date;
      opt.textContent = `${r.analysis_date} (${r.total_companies} companies, ${r.score_5_plus} scoring 5+)`;
      if (i === 0) opt.selected = true;
      datePicker.appendChild(opt);
    });

    // Auto-load the latest report
    await loadReport(reports[0].analysis_date);
  } catch (err) {
    datePicker.innerHTML = '<option value="">Error loading reports</option>';
    showStatus(statusEl, `Error loading report list: ${err.message}`, 'error');
  }
}

// ── Load and render a report ───────────────────────────────────────

async function loadReport(dateStr) {
  if (!dateStr) return;
  currentDate = dateStr;

  try {
    const data = dateStr === 'latest'
      ? await apiGet('/api/sec-leads/reports/latest')
      : await apiGet(`/api/sec-leads/reports/${dateStr}`);

    const stats = data.stats || {};
    const results = data.results || [];

    // Update stats bar
    filingsCount.textContent = stats.total_companies || 0;
    score5Count.textContent = stats.score_5_plus || 0;
    score7Count.textContent = stats.score_7_plus || 0;
    reportDate.textContent = stats.analysis_date || dateStr;
    statsGrid.classList.remove('hidden');

    // Filter to score >= 5 and render table
    const qualified = results.filter(r => r.score >= 5);
    qualified.sort((a, b) => b.score - a.score || a.company_name.localeCompare(b.company_name));

    renderTable(qualified);
    resultsCount.textContent = `${qualified.length} companies scoring 5+`;
    resultsArea.classList.remove('hidden');

  } catch (err) {
    showStatus(statusEl, `Error loading report: ${err.message}`, 'error');
  }
}

// ── Render table rows ──────────────────────────────────────────────

function renderTable(results) {
  if (!results.length) {
    tableBody.innerHTML = '<tr><td colspan="10" class="text-muted">No companies scored 5 or higher on this date.</td></tr>';
    return;
  }

  tableBody.innerHTML = results.map(r => {
    const color = scoreColor(r.score);

    // Determine primary contact
    let contactName = '';
    let contactTitle = '';
    let contactEmail = '';
    if (r.secretary_name) {
      contactName = r.secretary_name;
      contactTitle = r.secretary_title || 'Corporate Secretary';
      contactEmail = r.secretary_email || '';
    } else if (r.general_counsel_name) {
      contactName = r.general_counsel_name;
      contactTitle = r.general_counsel_title || 'General Counsel';
      contactEmail = r.general_counsel_email || '';
    } else if (r.board_chair_name) {
      contactName = r.board_chair_name;
      contactTitle = r.board_chair_title || 'Board Chair';
      contactEmail = r.board_chair_email || '';
    }

    // Parse board members for the new column
    let boardMembers = [];
    try {
      boardMembers = JSON.parse(r.board_members_json || '[]');
    } catch (_) { /* ignore */ }

    // Build board members cell with names and emails
    let boardHtml = '';
    if (boardMembers.length > 0) {
      boardHtml = boardMembers.map(m => {
        const name = escHtml(m.name || '');
        const title = m.title ? `<span class="sec-board-title">${escHtml(m.title)}</span>` : '';
        const email = m.email
          ? `<a href="mailto:${escHtml(m.email)}" class="sec-board-email">${escHtml(m.email)}</a>`
          : '';
        return `<div class="sec-board-member">${name}${title ? ' — ' + title : ''}${email ? '<br>' + email : ''}</div>`;
      }).join('');
    } else {
      boardHtml = '<span class="text-muted">—</span>';
    }

    const ticker = escHtml(r.ticker || '');
    const hasMemo = r.memo_text ? '' : ' disabled';
    const hasLetter = r.letter_text ? '' : ' disabled';

    return `<tr style="border-left: 4px solid ${color};">
      <td>${escHtml(r.analysis_date || '')}</td>
      <td>${escHtml(r.filing_date || '')}</td>
      <td><strong>${escHtml(r.company_name || '')}</strong><br><span class="sec-ticker">${ticker}</span></td>
      <td><a href="${escHtml(r.filing_url || '')}" target="_blank" rel="noopener">View 10-K on SEC</a></td>
      <td><span class="sec-score-badge" style="background:${color};">${r.score}</span></td>
      <td class="sec-gist-cell">${escHtml(r.gist || '')}</td>
      <td><strong>${escHtml(contactName)}</strong><br>${escHtml(contactTitle)}${contactEmail ? '<br><a href="mailto:' + escHtml(contactEmail) + '" class="sec-board-email">' + escHtml(contactEmail) + '</a>' : ''}</td>
      <td class="sec-board-cell">${boardHtml}</td>
      <td><button class="btn btn-primary btn-sm sec-memo-btn" data-ticker="${ticker}" data-date="${escHtml(r.analysis_date || '')}" data-company="${escHtml(r.company_name || '')}"${hasMemo}>Memo</button></td>
      <td><button class="btn-purple btn-sm sec-letter-btn" data-ticker="${ticker}" data-date="${escHtml(r.analysis_date || '')}" data-company="${escHtml(r.company_name || '')}"${hasLetter}>Letter</button></td>
    </tr>`;
  }).join('');

  // Enable sorting
  stampOriginalOrder(tableEl);
  enableTableSorting(tableEl);

  // Wire up memo/letter buttons
  tableBody.querySelectorAll('.sec-memo-btn').forEach(btn => {
    btn.addEventListener('click', () => {
      openDocument('memo', btn.dataset.date, btn.dataset.ticker, btn.dataset.company);
    });
  });
  tableBody.querySelectorAll('.sec-letter-btn').forEach(btn => {
    btn.addEventListener('click', () => {
      openDocument('letter', btn.dataset.date, btn.dataset.ticker, btn.dataset.company);
    });
  });
}

// ── Document modal ─────────────────────────────────────────────────

async function openDocument(type, date, ticker, companyName) {
  const label = type === 'memo' ? 'Memo' : 'Letter';
  modalTitle.textContent = `${label} — ${companyName}`;
  modalBody.textContent = 'Loading...';
  modal.classList.remove('hidden');

  try {
    const data = await apiGet(`/api/sec-leads/reports/${date}/${ticker}/${type}`);
    const text = type === 'memo' ? data.memo_text : data.letter_text;
    modalBody.textContent = text || `No ${type} available for this company.`;
  } catch (err) {
    modalBody.textContent = `Error loading ${type}: ${err.message}`;
  }
}

function closeModal() {
  modal.classList.add('hidden');
}

function copyToClipboard() {
  const text = modalBody.textContent;
  navigator.clipboard.writeText(text).then(() => {
    const orig = modalCopyBtn.textContent;
    modalCopyBtn.textContent = 'Copied!';
    setTimeout(() => { modalCopyBtn.textContent = orig; }, 2000);
  }).catch(() => {
    showStatus(statusEl, 'Failed to copy to clipboard', 'error');
  });
}

// ── Event listeners ────────────────────────────────────────────────

loadBtn.addEventListener('click', () => {
  const dateStr = datePicker.value;
  if (dateStr) loadReport(dateStr);
});

modalCloseBtn.addEventListener('click', closeModal);
modalBackdrop?.addEventListener('click', closeModal);
modalCopyBtn.addEventListener('click', copyToClipboard);

// Close modal with Escape key
document.addEventListener('keydown', e => {
  if (e.key === 'Escape' && !modal.classList.contains('hidden')) {
    closeModal();
  }
});

// ── Lazy load on first tab visit ───────────────────────────────────

document.querySelectorAll('.tab-btn').forEach(btn => {
  btn.addEventListener('click', () => {
    if (btn.dataset.tab === 'sec-leads' && !loaded) {
      loaded = true;
      loadReportDates();
    }
  });
});

File: api/routers/sec_leads.py
"""SEC Leads API endpoints — serve patent importance analysis results."""

from __future__ import annotations

from typing import Any, Dict

from fastapi import APIRouter, HTTPException
from google.cloud import bigquery

from api.config import settings
from api.services.bigquery_service import bq_service

router = APIRouter(prefix="/api/sec-leads", tags=["SEC Leads"])


@router.get("/reports")
def list_reports(limit: int = 30) -> Dict[str, Any]:
    """List available report dates with summary counts."""
    if limit > 100:
        limit = 100

    sql = f"""
    SELECT
      FORMAT_DATE('%Y-%m-%d', analysis_date) AS analysis_date,
      COUNT(*) AS total_companies,
      COUNTIF(score >= 5) AS score_5_plus,
      COUNTIF(score >= 7) AS score_7_plus,
      MAX(created_at) AS report_generated
    FROM `{settings.sec_leads_table}`
    GROUP BY analysis_date
    ORDER BY analysis_date DESC
    LIMIT @limit
    """
    params = [bigquery.ScalarQueryParameter("limit", "INT64", limit)]
    rows = bq_service.run_query(sql, params)

    reports = []
    for row in rows:
        reports.append({
            "analysis_date": row.get("analysis_date"),
            "total_companies": row.get("total_companies", 0),
            "score_5_plus": row.get("score_5_plus", 0),
            "score_7_plus": row.get("score_7_plus", 0),
            "report_generated": str(row.get("report_generated", "")),
        })

    return {"reports": reports}


@router.get("/reports/latest")
def get_latest_report() -> Dict[str, Any]:
    """Get all results for the most recent analysis date."""
    sql = f"""
    SELECT *
    FROM `{settings.sec_leads_table}`
    WHERE analysis_date = (
      SELECT MAX(analysis_date) FROM `{settings.sec_leads_table}`
    )
    ORDER BY score DESC, company_name ASC
    """
    rows = bq_service.run_query(sql)
    if not rows:
        return {"results": [], "stats": {}}

    return _format_report_response(rows)


@router.get("/reports/{report_date}")
def get_report(report_date: str) -> Dict[str, Any]:
    """Get all results for a specific analysis date."""
    sql = f"""
    SELECT *
    FROM `{settings.sec_leads_table}`
    WHERE analysis_date = @report_date
    ORDER BY score DESC, company_name ASC
    """
    params = [bigquery.ScalarQueryParameter("report_date", "DATE", report_date)]
    rows = bq_service.run_query(sql, params)
    if not rows:
        raise HTTPException(status_code=404, detail=f"No report found for {report_date}")

    return _format_report_response(rows)


@router.get("/reports/{report_date}/{ticker}/memo")
def get_memo(report_date: str, ticker: str) -> Dict[str, Any]:
    """Get the memo text for a specific company on a specific date."""
    sql = f"""
    SELECT memo_text, company_name
    FROM `{settings.sec_leads_table}`
    WHERE analysis_date = @report_date AND ticker = @ticker
    LIMIT 1
    """
    params = [
        bigquery.ScalarQueryParameter("report_date", "DATE", report_date),
        bigquery.ScalarQueryParameter("ticker", "STRING", ticker.upper()),
    ]
    rows = bq_service.run_query(sql, params)
    if not rows:
        raise HTTPException(status_code=404, detail="Memo not found")

    return {
        "company_name": rows[0].get("company_name", ""),
        "memo_text": rows[0].get("memo_text", ""),
    }


@router.get("/reports/{report_date}/{ticker}/letter")
def get_letter(report_date: str, ticker: str) -> Dict[str, Any]:
    """Get the letter text for a specific company on a specific date."""
    sql = f"""
    SELECT letter_text, company_name
    FROM `{settings.sec_leads_table}`
    WHERE analysis_date = @report_date AND ticker = @ticker
    LIMIT 1
    """
    params = [
        bigquery.ScalarQueryParameter("report_date", "DATE", report_date),
        bigquery.ScalarQueryParameter("ticker", "STRING", ticker.upper()),
    ]
    rows = bq_service.run_query(sql, params)
    if not rows:
        raise HTTPException(status_code=404, detail="Letter not found")

    return {
        "company_name": rows[0].get("company_name", ""),
        "letter_text": rows[0].get("letter_text", ""),
    }


def _format_report_response(rows: list) -> Dict[str, Any]:
    """Format BigQuery rows into the API response structure."""
    results = []
    for row in rows:
        results.append({
            "analysis_date": str(row.get("analysis_date", "")),
            "company_name": row.get("company_name", ""),
            "ticker": row.get("ticker", ""),
            "cik": row.get("cik", ""),
            "filing_date": str(row.get("filing_date", "")),
            "filing_url": row.get("filing_url", ""),
            "score": row.get("score", 0),
            "gist": row.get("gist", ""),
            "secretary_name": row.get("secretary_name"),
            "secretary_title": row.get("secretary_title"),
            "secretary_email": row.get("secretary_email"),
            "general_counsel_name": row.get("general_counsel_name"),
            "general_counsel_title": row.get("general_counsel_title"),
            "general_counsel_email": row.get("general_counsel_email"),
            "board_chair_name": row.get("board_chair_name"),
            "board_chair_title": row.get("board_chair_title"),
            "board_chair_email": row.get("board_chair_email"),
            "ceo_name": row.get("ceo_name"),
            "cfo_name": row.get("cfo_name"),
            "board_members_json": row.get("board_members_json", "[]"),
            "memo_text": row.get("memo_text"),
            "letter_text": row.get("letter_text"),
            "apollo_enriched": row.get("apollo_enriched", False),
        })

    # Compute stats
    total = len(results)
    score_5_plus = sum(1 for r in results if r["score"] >= 5)
    score_7_plus = sum(1 for r in results if r["score"] >= 7)
    analysis_date = results[0]["analysis_date"] if results else ""

    return {
        "stats": {
            "total_companies": total,
            "score_5_plus": score_5_plus,
            "score_7_plus": score_7_plus,
            "analysis_date": analysis_date,
        },
        "results": results,
    }
